managed/services/dump/dump.go

```diff
@@ -0,0 +1,299 @@
+// Copyright (C) 2023 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Package dump wraps pmm-dump integration.
+package dump
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"gopkg.in/reform.v1"
+
+	"github.com/percona/pmm/managed/models"
+)
+
+var ErrDumpAlreadyRunning = status.Error(codes.FailedPrecondition, "pmm-dump already running.")
+
+const (
+	pmmDumpBin = "pmm-dump"
+	dumpsDir   = "/srv/dump"
+)
+
+type Service struct {
+	l *logrus.Entry
+
+	db *reform.DB
+
+	running atomic.Bool
+
+	rw     sync.RWMutex
+	cancel context.CancelFunc
+}
+
+func New(db *reform.DB) *Service {
+	return &Service{
+		l:  logrus.WithField("component", "management/backup/backup"),
+		db: db,
+	}
+}
+
+type Params struct {
+	APIKey       string
+	Cookie       string
+	User         string
+	Password     string
+	ServiceNames []string
+	StartTime    *time.Time
+	EndTime      *time.Time
+	ExportQAN    bool
+	IgnoreLoad   bool
+}
+
+func (s *Service) StartDump(params *Params) (string, error) {
+	// Check if some pmm-dump already running.
+	if !s.running.CompareAndSwap(false, true) {
+		return "", ErrDumpAlreadyRunning
+	}
+
+	dump, err := models.CreateDump(s.db.Querier, models.CreateDumpParams{
+		ServiceNames: params.ServiceNames,
+		StartTime:    params.StartTime,
+		EndTime:      params.EndTime,
+		ExportQAN:    params.ExportQAN,
+		IgnoreLoad:   params.IgnoreLoad,
+	})
+	if err != nil {
+		s.running.Store(false)
+		return "", errors.Wrap(err, "failed to create dump")
+	}
+
+	l := s.l.WithField("dump_id", dump.ID)
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	s.rw.Lock()
+	s.cancel = cancel
+	s.rw.Unlock()
+
+	pmmDumpCmd := exec.CommandContext(ctx, //nolint:gosec
+		pmmDumpBin,
+		"export",
+		"--pmm-url=http://127.0.0.1",
```

Comment:
does this work with external VM and clickhouse?

Reply:
Nope, I will create ticket for that.